gpt4all-chat/src/database.cpp

```diff
@@ -1937,20 +2078,222 @@ QList<int> Database::searchEmbeddings(const std::vector<float> &query, const QLi
+        "a", "an", "and", "are", "as", "at", "be", "but", "by",
+        "for", "if", "in", "into", "is", "it", "no", "not", "of",
+        "on", "or", "such", "that", "the", "their", "then", "there",
+        "these", "they", "this", "to", "was", "will", "with"
+    };
+
+    QStringList rWords;
+    for (const QString &w : oWords)
+        if (!stopWords.contains(w.toLower()))
+            rWords << w;
+
+    QStringList quotedWords;
+    for (const QString &w : rWords)
+        quotedWords << "\"" + w + "\"";
+
+    BM25Query b;
+    b.type = 1; // broad search
+    b.input = oWords.join(" ");
+    b.query = "(" + quotedWords.join(" OR ") + ")";
+    b.qlength = 1; // length of phrase
+    b.ilength = oWords.size();
+    b.rlength = oWords.size() - rWords.size();
+    queries << b;
+    return queries;
+}
+
+QList<int> Database::searchBM25(const QString &query, const QList<QString> &collections, BM25Query &bm25q, int k)
+{
+    struct SearchResult { int chunkId; float dist; };
+    QList<BM25Query> bm25Queries = queriesForFTS5(query);
+
+    QList<SearchResult> results;
+    for (auto bm25Query : bm25Queries) {
+        QSqlQuery sqlQuery(m_db);
+        sqlQuery.prepare(SELECT_CHUNKS_FTS_SQL.arg(k));
+        sqlQuery.addBindValue(bm25Query.query);
+
+        if (!sqlQuery.exec()) {
+            qWarning() << "Database ERROR: Failed to execute BM25 query:" << sqlQuery.lastError();
+            return {};
+        }
+
+        if (!sqlQuery.next())
+            continue;
+
+        bm25q = bm25Query;
+
+        do {
+            const int chunkId = sqlQuery.value(0).toInt();
+            const float score = sqlQuery.value(1).toFloat();
+            results.append({chunkId, score});
+        } while (sqlQuery.next());
+        break;
+    }
+
+    k = qMin(k, results.size());
+    std::partial_sort(
+        results.begin(), results.begin() + k, results.end(),
+        [](const SearchResult &a, const SearchResult &b) { return a.dist < b.dist; }
+    );
+
+    QList<int> chunkIds;
+    chunkIds.reserve(k);
+    for (int i = 0; i < k; i++)
+        chunkIds << results[i].chunkId;
+    return chunkIds;
+}
+
+float Database::computeBM25Weight(const Database::BM25Query &bm25q)
+{
+    float bmWeight = 0.0f;
+    if (bm25q.type == 0 /*exact*/) {
+        bmWeight = 0.9f; // the highest we give
+    } else {
+        // qlength is the length of the phrases in the query by number of distinct words
+        // ilength is the length of the natural language query by number of distinct words
+        // rlength is the number of stop words removed from the natural language query to form the query
+        float queryLengthWeight = powf(bm25q.qlength, 2) / powf(float(bm25q.ilength - bm25q.rlength), 2);
+        queryLengthWeight = qBound(0.0f, queryLengthWeight, 1.0f);
+        bmWeight = queryLengthWeight;
+        bmWeight = qBound(0.0f, bmWeight, 1.0f);
+        bmWeight = (0.25f + bmWeight * (0.75f - 0.25f));
```

Comment:
Should add a comment briefly describing that this is an original formula and was produced experimentally. There are some simplifications that can be made here, such as `0.75f - 0.25f` -> `0.5f`, and `powf(bm25q.qlength, 2)` -> `1` as you mentioned.